### DISCUSSION TITLE
Do I need to worry about connection pool when passing tenantId in request using prisma client extension

### DISCUSSION THREAD

**Discussion by masterbater** [prisma/prisma#20553] [Q&A] [answered]
### Question

Basically once a tenant logged in and all request will be authenticated using jwt
There is a express middleware to decode jwt but I want also to append prisma client extension to request 
```typescript
const decoded = {
tenantId: 50
}
const xprisma = prisma.$extends({
  query: {
    $allModels: {
      $allOperations({ model, operation, args, query }) {
         args.where = { tenantId: decoded.tenantId, ...args.where }
        return query(args)
      },
    },
  },
})
request.prismaWithTenant = xprisma;
request.user = decoded
```
My question is does it create a connection pool everytime? 
Do I need to manage the connection and disconnection?

### Information about Prisma Schema, Client Queries and Environment (optional)

```js
// Add your schema.prisma and relevant query code
```

- OS: 
- Database: 
- Node.js version: 

- Run `prisma -v` to see your Prisma version and paste it

**Comment by ludralph** [comment, marked as answer]
Hi @masterbater 👋 

Thank you for raising this question.

Based on the code you provided, you are extending the Prisma client with an additional middleware that appends a `tenantId` to the `where` clause of each query based on the decoded JWT token. This extension will not create a new connection pool whenever a request is made. The connection pool is established when you create the Prisma client, and subsequent requests will reuse the same connection pool. See the [documentation](https://www.prisma.io/docs/concepts/components/prisma-client/working-with-prismaclient/connection-pool) on connection pool.

Prisma's connection pool efficiently manages and reuses connections to the database, allowing your application to handle multiple requests concurrently without creating a new connection for each request. This pooling mechanism helps in improving performance and resource utilization.

As long as you are creating the Prisma client outside the middleware, for example, when initializing your application or in some other central place, the connection pool will be shared among all requests, and you don't need to worry about managing connections and disconnections explicitly. However, it's essential to ensure that the Prisma client is initialized once and used as a shared instance across all requests in your Express application. Creating a new Prisma client instance for each request could lead to resource exhaustion and poor performance.

So, in summary, you don't need to manage the connection and disconnection explicitly when using Prisma with Express, as long as you initialize the Prisma client only once and use it as a shared instance throughout the application. The connection pooling mechanism will handle the efficient management of database connections for you.

If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Reply by masterbater** [reply]
So you mean as long as I just extend the singleton prisma it doesnt create new connection.

**Reply by ludralph** [reply]
Yes, that's correct. As long as you extend the singleton Prisma client instance (which is created once and reused throughout your application), it won't create a new database connection each time you extend it. The connection pooling mechanism will handle reusing and managing the database connections efficiently.

**Reply by masterbater** [reply]
If I create multiple extend is it immutable or mutable?

**Reply by ankur-arch** [reply]
@masterbater extending the Client itself is an immutable operation as it returns a new copy rather than mutating the original instance directly. 
```typescript
const prisma = new PrismaClient()
const xPrimsa = prisma.$extends() // returns a new instance of Prisma 
```
